### DISCUSSION TITLE
3.4.3: Maintenance release

### DISCUSSION THREAD

**Discussion by guyer** [usnistgov/fipy#873] [Announcements]
This maintenance release adds a new example contributed by @Jon83Carvalho, clarifies many points in the documentation, migrates all CI to [Azure](https://dev.azure.com), updates to using [wheels](https://packaging.python.org/en/latest/specifications/binary-distribution-format/) for distribution, and substantially refactors matrices to work more consistently across solvers.

## What's Changed
* Debug CI failures by @guyer in https://github.com/usnistgov/fipy/pull/749
* Stokes Cavity - non-newtonian by @Jon83Carvalho in https://github.com/usnistgov/fipy/pull/748
* Disable TVTK tests if its prerequisites aren't met by @guyer in https://github.com/usnistgov/fipy/pull/764
* Tabulate versions of FiPy dependencies when tests are run by @guyer in https://github.com/usnistgov/fipy/pull/763
* Refactor matrices by @guyer in https://github.com/usnistgov/fipy/pull/721
* Raise ImportError before trying to unpack solvers by @guyer in https://github.com/usnistgov/fipy/pull/768
* Introduce SharedTemporaryFile (bis) by @guyer in https://github.com/usnistgov/fipy/pull/769
* Get website documentation to display math by @guyer in https://github.com/usnistgov/fipy/pull/773
* Add future package to Binder environment by @guyer in https://github.com/usnistgov/fipy/pull/779
* Document parallel gmsh by @guyer in https://github.com/usnistgov/fipy/pull/793
* Remove future from setup.py and _setup commands by @guyer in https://github.com/usnistgov/fipy/pull/785
* Fix broken circleSquare with newer Gmsh by @guyer in https://github.com/usnistgov/fipy/pull/794
* Move trilinos tests to Py3k by @guyer in https://github.com/usnistgov/fipy/pull/797
* Replace deprecated numpy types by @guyer in https://github.com/usnistgov/fipy/pull/798
* Fix Python 2.7 conda environment by @wd15 in https://github.com/usnistgov/fipy/pull/795
* Speed up conda environment creation by @guyer in https://github.com/usnistgov/fipy/pull/804
* Fix python 3.9 represenation of abs by @guyer in https://github.com/usnistgov/fipy/pull/812
* Set up CI with Azure Pipelines by @guyer in https://github.com/usnistgov/fipy/pull/822
* fix: stop divide by zero warning in LU solvers by @wd15 in https://github.com/usnistgov/fipy/pull/790
* Change `colorbar` argument to valid value by @guyer in https://github.com/usnistgov/fipy/pull/829
* Get gmsh version from stderr or stdout by @guyer in https://github.com/usnistgov/fipy/pull/846
* Collect contact information on single page by @guyer in https://github.com/usnistgov/fipy/pull/847
* Debug CIs by @guyer in https://github.com/usnistgov/fipy/pull/848
* Refine azure runs by @guyer in https://github.com/usnistgov/fipy/pull/851
* Issue858 pysparse ci failure by @guyer in https://github.com/usnistgov/fipy/pull/859
* Fix missing Qt on headless CI by @guyer in https://github.com/usnistgov/fipy/pull/864
* Update CI documentation to refer only to Azure by @guyer in https://github.com/usnistgov/fipy/pull/863
* Shift CI from eggs to wheels by @guyer in https://github.com/usnistgov/fipy/pull/860
* Document the things by @guyer in https://github.com/usnistgov/fipy/pull/866
* Document viewer attributes by @guyer in https://github.com/usnistgov/fipy/pull/867
* Define _localNonOverlappingFaceIDs for Gmsh by @guyer in https://github.com/usnistgov/fipy/pull/857
* Update changelog by @guyer in https://github.com/usnistgov/fipy/pull/871

## New Contributors
* @Jon83Carvalho made their first contribution in https://github.com/usnistgov/fipy/pull/748

**Curated Changelog**: https://github.com/usnistgov/fipy/blob/3.4.3/CHANGELOG.rst

**Full Changelog**: https://github.com/usnistgov/fipy/compare/3.4.2.1...3.4.3

<hr /><em>This discussion was created from the release <a href='https://github.com/usnistgov/fipy/releases/tag/3.4.3'>3.4.3: Maintenance release</a>.</em>

**Comment by mhvwerts** [comment]
Dear @guyer and colleagues,

Many thanks for Fipy 3.4.3! What a coincidence, we are currently working with our IT department to prepare the computer rooms of our school for next year's courses, and evaluating different options for our Python stack.

We unadvertedly ran the new version on JupyterLite (with Python 3.10), not knowing that 3.4.3 had just appeared. JupyterLite is a Jupyter Notebook that runs completely in the web browser using Javascript/Webassembly technology. This is of interest for teaching purposes, as the students do not have to install any specific software stack. It might even run on a smartphone.

The good news is that we were able to use JupyterLite and run Fipy on a simple cylindrical 1D diffusion problem, all inside a web browser. We did not test any further and will probably continue using a local conda installation instead of JupyterLite for our course. 

A JupyterLite instance can be opened via: https://jupyterlite.readthedocs.io/en/latest/_static/lab/index.html

Create a Notebook. In the first cell of the Notebook, put the following code.

```python
import piplite
await piplite.install("fipy")
```

and in the next cell:

```python
import fipy as fp
print('Fipy version',fp.__version__)
```

The `fipy` import throws some warnings, probably related to Python 3.10 around which the present JupyterLite is centered. Finally, it prints `Fipy version 3.4.3`

Our [example/test notebook](https://github.com/mhvwerts/Python-mecatronique/blob/master/Example%202%20-%20Scipy%20special%20functions%20and%20Fipy.ipynb) worked fine with JupyterLite (with the `import piplite; await piplite.install("fipy")` inserted as the first cell).

Thanks again for Fipy.

**Comment by guyer** [comment]
That's great to hear, @mhvwerts. The warnings are probably nothing to worry about, but please share if you're at all concerned.

**Reply by mhvwerts** [reply]
Indeed, nothing to worry about, just `SyntaxWarning`s about `is` and `is not`.

Here is the complete list of warnings generated by `import fipy` when in JupyterLite/Python 3.10.

```
/lib/python3.10/site-packages/fipy/variables/variable.py:649: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  if v.shape is not ():
/lib/python3.10/site-packages/fipy/variables/meshVariable.py:118: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  if valueShape is not () and valueShape[-1] == self._globalNumberOfElements:
/lib/python3.10/site-packages/fipy/variables/meshVariable.py:523: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  and otherShape is not ()
/lib/python3.10/site-packages/fipy/meshes/abstractMesh.py:1148: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if arr.dtype.name is 'bool':
/lib/python3.10/site-packages/fipy/meshes/builders/utilityClasses.py:62: SyntaxWarning: "is not" with a literal. Did you mean "!="?
  if numerix.getShape(d) is not ():
/lib/python3.10/site-packages/fipy/terms/abstractConvectionTerm.py:155: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if (coeffShape is ()) or (coeffShape[0] != var.mesh.dim):
```

**Reply by mhvwerts** [reply]
It looks a bit like issue #729 .

**Reply by guyer** [reply]
Yep, the same thing for sure

**Comment by guyer** [comment]
FYI: [anaconda](https://github.com/conda/infra/issues/561) and/or [conda-forge](https://github.com/conda-forge/conda-forge.github.io/issues/1779) are having a major breakage right now, leading to 3.4.3 not being available via `conda`, yet.

**Reply by guyer** [reply]
Breakage resolved
